### PR TITLE
[SPARK-10618] [Mesos] Refactoring scheduling condition and adding test for same

### PR DESCRIPTION
**Pull request by SleepyThread**
This commit just refactored a scheduling condition into the method and added test for the same.

### CODE REVIEW

**Comment by SleepyThread** [comment]
@andrewor14 Can you please take a look at it?

File: core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala [hunk 1]
```diff
@@ -184,4 +184,52 @@ class CoarseMesosSchedulerBackendSuite extends SparkFunSuite
...
+      .set("spark.cores.max", "10")
+
+    val sc = new SparkContext(sparkConf)
+
+    val driver = mock[SchedulerDriver]
+    when(driver.start()).thenReturn(Protos.Status.DRIVER_RUNNING)
+    val taskScheduler = mock[TaskSchedulerImpl]
+    when(taskScheduler.sc).thenReturn(sc)
+
+
```
Comment (tnachen) [inline core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala]: kill space

File: core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala [hunk 2]
```diff
@@ -184,4 +184,52 @@ class CoarseMesosSchedulerBackendSuite extends SparkFunSuite
...
+    assert(schedulerBackend.isOfferValidForScheduling(true, "Slave2", 10000, 5, sc) === false)
+
+    // Return false When task is failed more than MAX_SLAVE_FAILURES times on the given slave
+    assert(schedulerBackend.isOfferValidForScheduling(true, "Slave3", 10000, 5, sc) === false)
+
+    schedulerBackend.totalCoresAcquired = 10
+
+    // Return false When max core is already acquired
+    assert(schedulerBackend.isOfferValidForScheduling(true, "Slave1", 10000, 5, sc) === false)
+
```
Comment (tnachen) [inline core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala]: kill space

File: core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala [hunk 3]
```diff
@@ -184,4 +184,52 @@ class CoarseMesosSchedulerBackendSuite extends SparkFunSuite
...
+    // Return false When offer do not meet constraints
+    assert(schedulerBackend.isOfferValidForScheduling(false, "Slave1", 10000, 5, sc) === false)
+
+    // Return false When memory in offer is less than required memory
+    assert(schedulerBackend.isOfferValidForScheduling(true, "Slave1", 1, 5, sc) === false)
+
+    // Return false When cpu in offer is less than required cpu
+    assert(schedulerBackend.isOfferValidForScheduling(true, "Slave1", 10000, 0, sc) === false)
+
```
Comment (tnachen) [inline core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala]: I think we can remove the spaces here

File: core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala [hunk 4]
```diff
@@ -184,4 +184,52 @@ class CoarseMesosSchedulerBackendSuite extends SparkFunSuite
 
     verify(driver, times(1)).reviveOffers()
   }
+
+  test("testing various isOfferValidForScheduling cases") {
```
Comment (tnachen) [inline core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackendSuite.scala]: Ideally I would like to see each test case tests one condition as it's hard to know what's going on here.
Can you at least organized them to related ones (constraints, required mem/cpu, etc).

**Comment by tnachen** [comment]
I think this is a good idea, can you do this for fine grain mode too?

**Comment by dragos** [comment]
Not a bad idea, though I don't think this should go in before #8639, which it conflicts with (and which adds real functionality, and is reviewed).

**Comment by SleepyThread** [comment]
I agree with @dragos. After #8639 is merged in master. I will update this pull request. 

@tnachen I am a fan of small changes and hence will it be ok if i open another pull request for fine-grained mode ?

**Comment by SleepyThread** [comment]
Refactored and added test for fine grained mode too. Waiting for #8639 to be merged in master.

**Comment by tnachen** [comment]
Actually I'm thinking what could be better is that if we can log what's the exact condition that didn't pass which made us skip the offer. It becomes a bit hard for users to guess why no jobs are running.
How about making each testing condition a rule, and whenever any of them wasn't matched we also have a corresponding message saying why the offer was rejected with the exact reason?

**Comment by SleepyThread** [comment]
@tnachen Its really a great idea. But we should log these things in debug mode, rather than info log otherwise it will screw up the logs.

**Comment by showroomtv** [comment]
Yes please log debug :) But at least the information is available.

**Comment by tnachen** [comment]
Yes please log debug :) But at least the information is available.

**Comment by andrewor14** [comment]
@SleepyThread #8639 is merged into master now. Can you update this?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 5]
```diff
@@ -244,13 +244,7 @@ private[spark] class CoarseMesosSchedulerBackend(
...
         val cpus = getResource(offer.getResourcesList, "cpus").toInt
         val id = offer.getId.getValue
-        if (taskIdToSlaveId.size < executorLimit &&
-            totalCoresAcquired < maxCores &&
-            meetsConstraints &&
-            mem >= calculateTotalMemory(sc) &&
-            cpus >= 1 &&
-            failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
-            !slaveIdsWithExecutors.contains(slaveId)) {
+        if (isOfferValidForScheduling(meetsConstraints, slaveId, mem, cpus, sc)) {
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: I think the wording is a bit odd, it's not that the offer is not valid, it usually doesn't meet the job requirements.
How about naming this isOfferSatisfiesRequirements?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 6]
```diff
@@ -292,6 +286,18 @@ private[spark] class CoarseMesosSchedulerBackend(
   }
 
 
+  def isOfferValidForScheduling(meetsConstraints: Boolean,
+                                slaveId: String, mem: Double,
+                                cpus: Int, sc: SparkContext): Boolean = {
+    taskIdToSlaveId.size < executorLimit &&
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Let's leave a TODO here that we want to abstract out each condition to have a specific error message logged why a offer doesn't meet the requirement.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 7]
```diff
@@ -219,12 +219,9 @@ private[spark] class MesosSchedulerBackend(
         //  2. Memory requirements
         //  3. CPU requirements - need at least 1 for executor, 1 for task
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: You should move these comments to the new method you have.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 8]
```diff
@@ -301,6 +298,15 @@ private[spark] class MesosSchedulerBackend(
     }
   }
 
+  def isOfferValidForScheduling(cpusOffered: Double, memory : Double,
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: Same naming suggestion

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 9]
```diff
@@ -301,6 +298,15 @@ private[spark] class MesosSchedulerBackend(
...
+                                slaveId: String, sc : SparkContext): Boolean = {
+    val meetsMemoryRequirements = memory >= calculateTotalMemory(sc)
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: Why do you have these variables in fine grain but not in coarse grain?
Let's make them consisitent

**Comment by tnachen** [comment]
Just took another pass of this review, besides the naming suggestions and style fixes overall it LGTM.
Also please update the title of this PR since it's updating both coarse grain and fine grain mode.

**Comment by andrewor14** [comment]
@SleepyThread have you had the chance to address the comments? Are you still working on this?

**Comment by SleepyThread** [comment]
@andrewor14 I have addressed all the comments. Please review the commit.

**Comment by andrewor14** [comment]
Can you rebase to master? Then @dragos or @tnachen can do another pass.

**Comment by SleepyThread** [comment]
@andrewor14 @dragos @tnachen i screwed up the merge. I will open up a new pull request which is sync with master.

**Comment by SleepyThread** [comment]
Please refer new pull request #10326. Closing this pull request.

**Pull request closed without merging**